java/core/src/test/java/org/bondlib/StructGenerator.java

```diff
@@ -0,0 +1,408 @@
+            BondType setElementBondType = setBondType.getElementType();
+            Set setValue = setBondType.newDefaultValue();
+            this.generateCollectionElements(setValue, setElementBondType, context);
+            returnValue = setValue;
+        } else if (bondTypeClass == MapBondType.class) {
+            // map<,>
+            MapBondType mapBondType = (MapBondType) bondType;
+            BondType mapKeyBondType = mapBondType.getKeyType();
+            BondType mapValueBondType = mapBondType.getValueType();
+            Map mapValue = mapBondType.newDefaultValue();
+            this.generateMapElements(mapValue, mapKeyBondType, mapValueBondType, context);
+            returnValue = mapValue;
+        } else if (bondTypeClass == BondedBondType.class) {
+            // bonded<>
+            BondedBondType bondedBondType = (BondedBondType) bondType;
+            StructBondType declaredStructBondType = bondedBondType.getValueType();
+            ArrayList<StructBondType> allowedPolymorphicStructBondTypes =
+                    this.getPolymorphicStructBondTypesFor(declaredStructBondType);
+            int index = this.rand.nextInt(allowedPolymorphicStructBondTypes.size());
+            StructBondType actualStructBondType = allowedPolymorphicStructBondTypes.get(index);
+            BondSerializable structValue = this.generateStruct(actualStructBondType, context);
+            returnValue = Bonded.fromObject(structValue);
+        } else if (EnumBondType.class.isAssignableFrom(bondTypeClass)) {
+            // generated enum
+            EnumBondType enumBondType = (EnumBondType) bondType;
+            returnValue = this.generateEnumValue(enumBondType);
+        } else if (StructBondType.class.isAssignableFrom(bondTypeClass)) {
+            StructBondType structBondType = (StructBondType) bondType;
+            returnValue = this.generateStruct(structBondType, context);
+        } else {
+            // unknown
+            throw new IllegalArgumentException("Unknown BondType implementation: " + bondTypeClass);
+        }
+        return returnValue;
+    }
+
+    // returns true if a pseudo-random number distributed uniformly between 0 and 1 is less than the argument
+    private boolean testProbability(double probability) {
+        return this.rand.nextDouble() < probability;
+    }
+
+    // helper to generate a pseudo-random integer with the given number of bits (optionally signed)
+    private long generateIntValue(int bitCount, boolean isSigned) {
+        // Given number N randomly chosen between 1 and bitCount with uniform probability, determine the
+        // magnitude of the result as an unsigned value of N bits where bits 0 to N-1 are randomly chosen
+        // individually (with uniform probability), and all the other bits are set to 0. Then optionally
+        // negate the result with 50% probability.
+        // This algorithm improves coverage of smaller integers for variable-length encoding.
```

Comment:
I agree with this algorithm, but this comment didn't help much with getting there. How about putting the fact that it isn't uniform and the justification for that up front?
`
We want uniform distribution over magnitudes of numbers, rather than values, to exercise smaller varints as frequently as larger ones. Therefore, we first pick a random length up to bitCount, then a random number that fits in that number of bits. If isSigned, we negate that number with 50% probability. 
`

Additionally, because this is critical to understanding the meaning and use of `generateIntValue`, this should be a Javadoc. It doesn't matter that it's private - making it a Javadoc will allow other devs to see it in tooltips or whatever the IDE has instead of having to leave the callsite.

Reply:
Made this method (and a couple others) public and added javadoc.

---
In reply to: [138758973](https://github.com/Microsoft/bond/pull/607#discussion_r138758973) [](ancestors = 138758973)